cmd/lifecycle/builder.go

```diff
@@ -48,26 +60,160 @@ func (b *buildCmd) Args(nargs int, args []string) error {
 }
 
 func (b *buildCmd) Privileges() error {
-	// builder should never be run with privileges
-	if priv.IsPrivileged() {
+	_, privGroup, _, err := b.readData()
+	if err != nil {
+		return err
+	}
+
+	hasPrivGroups := len(privGroup.Group) > 0
+
+	// builder should never be run with privileges if there aren't any stack buildpacks
+	if !hasPrivGroups && priv.IsPrivileged() {
 		return cmd.FailErr(errors.New("refusing to run as root"), "build")
+	} else if hasPrivGroups && !priv.IsPrivileged() {
+		return cmd.FailErr(errors.New("must run as root"), "build")
 	}
 	return nil
 }
 
 func (b *buildCmd) Exec() error {
-	group, plan, err := b.readData()
+	group, privGroup, plan, err := b.readData()
+	if err != nil {
+		return err
+	}
+
+	if err := verifyBuildpackApis(group, privGroup); err != nil {
+		return err
+	}
+
+	if len(privGroup.Group) == 0 {
+		return b.build(group, plan)
+	}
+	return b.buildWithReexec(group, privGroup, plan)
+}
+
+func (ba buildArgs) buildAll(group, privGroup lifecycle.BuildpackGroup, plan lifecycle.BuildPlan, rootuid, rootgid, uid, gid int) error {
+	if len(privGroup.Group) > 0 {
+		if err := priv.RunAsEffective(rootuid, rootgid); err != nil {
```

Comment:
Didn't we already drop privileges by this point when running the creator?

Reply:
Not anymore. The creator was updated to use `RunAsEffective`, which now allows us to regain privileges after dropping them. I did this work prior to discussing creator can't really ever fully do stackpacks...but it CAN do stackpacks as long as they are `build:` phase only requirements.

Reply:
Ohhh I missed that part. Thanks for explaining :)